bemani/backend/ddr/ddrsn2/ddrsn2.py

```diff
@@ -0,0 +1,566 @@
+            self.GAME_CHART_SINGLE_DIFFICULT: self.CHART_SINGLE_DIFFICULT,
+            self.GAME_CHART_SINGLE_EXPERT: self.CHART_SINGLE_EXPERT,
+            self.GAME_CHART_SINGLE_CHALLENGE: self.CHART_SINGLE_CHALLENGE,
+            self.GAME_CHART_DOUBLE_BASIC: self.CHART_DOUBLE_BASIC,
+            self.GAME_CHART_DOUBLE_DIFFICULT: self.CHART_DOUBLE_DIFFICULT,
+            self.GAME_CHART_DOUBLE_EXPERT: self.CHART_DOUBLE_EXPERT,
+            self.GAME_CHART_DOUBLE_CHALLENGE: self.CHART_DOUBLE_CHALLENGE,
+        }[game_chart]
+
+    def db_to_game_chart(self, db_chart: int) -> int:
+        return {
+            self.CHART_SINGLE_BEGINNER: self.GAME_CHART_SINGLE_BEGINNER,
+            self.CHART_SINGLE_BASIC: self.GAME_CHART_SINGLE_BASIC,
+            self.CHART_SINGLE_DIFFICULT: self.GAME_CHART_SINGLE_DIFFICULT,
+            self.CHART_SINGLE_EXPERT: self.GAME_CHART_SINGLE_EXPERT,
+            self.CHART_SINGLE_CHALLENGE: self.GAME_CHART_SINGLE_CHALLENGE,
+            self.CHART_DOUBLE_BASIC: self.GAME_CHART_DOUBLE_BASIC,
+            self.CHART_DOUBLE_DIFFICULT: self.GAME_CHART_DOUBLE_DIFFICULT,
+            self.CHART_DOUBLE_EXPERT: self.GAME_CHART_DOUBLE_EXPERT,
+            self.CHART_DOUBLE_CHALLENGE: self.GAME_CHART_DOUBLE_CHALLENGE,
+        }[db_chart]
+
+    def db_to_game_halo(self, db_halo: int) -> int:
+        if db_halo == self.HALO_MARVELOUS_FULL_COMBO:
+            combo_type = self.GAME_HALO_MARVELOUS_COMBO
+        elif db_halo == self.HALO_PERFECT_FULL_COMBO:
+            combo_type = self.GAME_HALO_PERFECT_COMBO
+        elif db_halo == self.HALO_GREAT_FULL_COMBO:
+            combo_type = self.GAME_HALO_FULL_COMBO
+        else:
+            combo_type = self.GAME_HALO_NONE
+        return combo_type
+
+    def sn2_to_db_halo(self, full: int, perf: int) -> int:
+        if full == 1 and perf == 1:
+            combo_type = self.HALO_PERFECT_FULL_COMBO
+        elif full == 1:
+            combo_type = self.HALO_GREAT_FULL_COMBO
+        else:
+            combo_type = self.HALO_NONE
+        return combo_type
+
+    def handle_game_common_request(self, request: Node) -> Node:
+        game = Node.void("game")
+        for flagid in range(256):
+            flag = Node.void("flag")
+            game.add_child(flag)
+
+            flag.set_attribute("id", str(flagid))
+            flag.set_attribute("s2", "0")
+            flag.set_attribute("s1", "0")
+            flag.set_attribute("t", "0")
+
+        hit_chart = self.data.local.music.get_hit_chart(self.game, self.music_version, self.GAME_MAX_SONGS)
+        counts_by_reflink = [0] * self.GAME_MAX_SONGS
+        for reflink, plays in hit_chart:
+            if reflink >= 0 and reflink < self.GAME_MAX_SONGS:
+                counts_by_reflink[reflink] = plays
+        game.add_child(Node.u32_array("cnt_music", counts_by_reflink))
+
+        return game
+
+    def handle_game_hiscore_request(self, request: Node) -> Node:
```

Comment:
A lot of this stuff is most likely unused, I used the ddrx code as a baseline, need to take a look

Reply:
Yeah, I would check whether these are even in the game at all, I would doubt it.